File: bot_actions.py
import asyncio
import logging
import math
from typing import Dict, Tuple

logger = logging.getLogger(__name__)

class BotActions:
    def __init__(self, connection):
        self.connection = connection
        self.current_task = None
        self.following = None
        
        self.block_types = {
            'wood': ['oak_log', 'birch_log', 'spruce_log', 'jungle_log'],
            'madeira': ['oak_log', 'birch_log', 'spruce_log', 'jungle_log'],
            'stone': ['stone', 'cobblestone'],
            'pedra': ['stone', 'cobblestone'],
            'dirt': ['dirt'],
            'terra': ['dirt']
        }
    
    async def goto_player(self, player_name: str):
        logger.info(f"Indo até o jogador {player_name}")
        
        if self.current_task:
            self.current_task.cancel()
        
        self.current_task = asyncio.create_task(self._goto_player_task(player_name))
        await self.current_task
    
    async def _goto_player_task(self, player_name: str):
        try:
            for _ in range(100):
                bot_pos = self.connection.get_position()
                
                target_pos = await self._find_player_position(player_name)
                
                if not target_pos:
                    await asyncio.sleep(0.5)
                    continue
                
                distance = self._calculate_distance(bot_pos, target_pos)
                
                if distance < 2.0:
                    logger.info(f"Chegou ao jogador {player_name}")
                    await self.connection.send_chat(f"Cheguei até você, {player_name}!")
                    break
                
                next_pos = self._calculate_next_position(bot_pos, target_pos)
                await self.connection.move_to(next_pos['x'], next_pos['y'], next_pos['z'])
                
                await asyncio.sleep(0.2)
            
        except asyncio.CancelledError:
            logger.info("Tarefa de movimento cancelada")
        except Exception as e:
            logger.error(f"Erro ao ir até jogador: {e}")
    
    async def follow_player(self, player_name: str):
        logger.info(f"Seguindo jogador {player_name}")
        
        if self.current_task:
            self.current_task.cancel()
        
        self.following = player_name
        self.current_task = asyncio.create_task(self._follow_player_task(player_name))
    
    async def _follow_player_task(self, player_name: str):
        try:
            while self.following == player_name:
                bot_pos = self.connection.get_position()
                target_pos = await self._find_player_position(player_name)
                
                if not target_pos:
                    await asyncio.sleep(0.5)
                    continue
                
                distance = self._calculate_distance(bot_pos, target_pos)
                
                if distance > 3.0:
                    next_pos = self._calculate_next_position(bot_pos, target_pos)
                    await self.connection.move_to(next_pos['x'], next_pos['y'], next_pos['z'])
                
                await asyncio.sleep(0.3)
                
        except asyncio.CancelledError:
            logger.info("Tarefa de seguir cancelada")
        except Exception as e:
            logger.error(f"Erro ao seguir jogador: {e}")
    
    async def collect_blocks(self, block_type: str, quantity: int):
        logger.info(f"Coletando {quantity} {block_type}")
        
        if self.current_task:
            self.current_task.cancel()
        
        self.current_task = asyncio.create_task(self._collect_blocks_task(block_type, quantity))
        await self.current_task
    
    async def _collect_blocks_task(self, block_type: str, quantity: int):
        try:
            blocks_to_find = self.block_types.get(block_type.lower(), [block_type])
            
            await self.connection.send_chat(f"Procurando {block_type}...")
            
            collected = 0
            search_radius = 50
            
            bot_pos = self.connection.get_position()
            
            for x_offset in range(-search_radius, search_radius, 5):
                for z_offset in range(-search_radius, search_radius, 5):
                    if collected >= quantity:
                        break
                    
                    target_x = bot_pos['x'] + x_offset
                    target_z = bot_pos['z'] + z_offset
                    target_y = bot_pos['y']
                    
                    await self.connection.move_to(target_x, target_y, target_z)
                    await asyncio.sleep(0.3)
                    
                    for y_offset in range(-2, 3):
                        block_x = int(target_x)
                        block_y = int(target_y + y_offset)
                        block_z = int(target_z)
                        
                        await self.connection.break_block(block_x, block_y, block_z)
                        collected += 1
                        
                        if collected >= quantity:
                            break
                        
                        await asyncio.sleep(0.1)
                
                if collected >= quantity:
                    break
            
            logger.info(f"Coletou {collected} blocos de {block_type}")
            await self.connection.send_chat(f"Coletei {collected} {block_type}!")
            
        except asyncio.CancelledError:
            logger.info("Tarefa de coleta cancelada")
        except Exception as e:
            logger.error(f"Erro ao coletar blocos: {e}")
    
    async def give_items(self, player_name: str, item_type: str, quantity: int):
        logger.info(f"Entregando {quantity} {item_type} para {player_name}")
        
        await self.connection.send_command(f"give {player_name} {item_type} {quantity}")
        
        await asyncio.sleep(0.5)
    
    async def stop_all(self):
        logger.info("Parando todas as ações")
        
        self.following = None
        
        if self.current_task:
            self.current_task.cancel()
            self.current_task = None
        
        await self.connection.send_chat("Parei todas as ações!")
    
    async def _find_player_position(self, player_name: str) -> Dict:
        positions = {
            player_name: {
                'x': self.connection.position['x'] + 10,
                'y': self.connection.position['y'],
                'z': self.connection.position['z'] + 10
            }
        }
        
        return positions.get(player_name)
    
    def _calculate_distance(self, pos1: Dict, pos2: Dict) -> float:
        dx = pos2['x'] - pos1['x']
        dy = pos2['y'] - pos1['y']
        dz = pos2['z'] - pos1['z']
        
        return math.sqrt(dx*dx + dy*dy + dz*dz)
    
    def _calculate_next_position(self, current: Dict, target: Dict, speed: float = 1.0) -> Dict:
        dx = target['x'] - current['x']
        dy = target['y'] - current['y']
        dz = target['z'] - current['z']
        
        distance = math.sqrt(dx*dx + dy*dy + dz*dz)
        
        if distance < speed:
            return target
        
        ratio = speed / distance
        
        return {
            'x': current['x'] + dx * ratio,
            'y': current['y'] + dy * ratio,
            'z': current['z'] + dz * ratio
        }

File: botActions.js
export class BotActions {
  constructor(bot, pathfinder) {
    this.bot = bot;
    this.pathfinder = pathfinder;
    this.miningQueue = [];
    this.inventory = new Map();
  }

  async executeAction(actionData, playerName, playerPosition) {
    const { action, target, quantity, response } = actionData;

    this.sendMessage(response);

    switch (action) {
      case 'goto':
        await this.goToPlayer(playerPosition);
        break;

      case 'mine':
        await this.mineResource(target, quantity || 1);
        break;

      case 'give':
        await this.goToPlayer(playerPosition);
        await this.giveItems(playerName, target, quantity);
        break;

      case 'follow':
        this.pathfinder.startFollowing(playerName);
        break;

      case 'stop':
        this.pathfinder.stopFollowing();
        break;

      case 'respond':
        break;

      default:
        console.log(`Ação desconhecida: ${action}`);
    }
  }

  async goToPlayer(playerPosition) {
    if (!playerPosition) {
      this.sendMessage('Não consigo ver sua posição!');
      return;
    }

    console.log(`Indo até o jogador em: ${JSON.stringify(playerPosition)}`);
    await this.pathfinder.goToPosition(playerPosition, 2);
    this.sendMessage('Cheguei!');
  }

  async mineResource(resourceType, quantity) {
    console.log(`Tentando minerar ${quantity}x ${resourceType}`);
    
    this.sendMessage(`Procurando por ${resourceType}...`);
    
    const blockTypes = {
      'oak_log': ['oak_log', 'log', 'wood'],
      'madeira': ['oak_log', 'log', 'wood', 'birch_log', 'spruce_log'],
      'pedra': ['stone', 'cobblestone'],
      'stone': ['stone', 'cobblestone'],
      'coal': ['coal_ore'],
      'iron': ['iron_ore'],
      'carvao': ['coal_ore'],
      'ferro': ['iron_ore']
    };

    const searchBlocks = blockTypes[resourceType] || [resourceType];
    
    for (let i = 0; i < quantity; i++) {
      const block = await this.findNearbyBlock(searchBlocks, 50);
      
      if (block) {
        await this.pathfinder.goToPosition(block.position, 4);
        await this.mineBlock(block.position);
        this.addToInventory(resourceType, 1);
        
        if (i < quantity - 1) {
          await this.sleep(1000);
        }
      } else {
        this.sendMessage(`Só consegui encontrar ${i} ${resourceType}. Não há mais por perto.`);
        break;
      }
    }

    this.sendMessage(`Coletei ${Math.min(quantity, this.inventory.get(resourceType) || 0)} ${resourceType}!`);
  }

  async findNearbyBlock(blockTypes, radius) {
    if (!this.bot.position) return null;

    const botPos = this.bot.position;
    let closestBlock = null;
    let closestDistance = Infinity;

    for (let x = -radius; x <= radius; x++) {
      for (let y = -radius; y <= radius; y++) {
        for (let z = -radius; z <= radius; z++) {
          const pos = {
            x: Math.floor(botPos.x) + x,
            y: Math.floor(botPos.y) + y,
            z: Math.floor(botPos.z) + z
          };

          const distance = Math.sqrt(x * x + y * y + z * z);
          
          if (distance < closestDistance && distance <= radius) {
            for (const blockType of blockTypes) {
              if (this.isBlockType(pos, blockType)) {
                closestBlock = { position: pos, type: blockType };
                closestDistance = distance;
                break;
              }
            }
          }
        }
      }
    }

    return closestBlock;
  }

  isBlockType(position, blockType) {
    return Math.random() > 0.95;
  }

  async mineBlock(position) {
    console.log(`Minerando bloco em: ${JSON.stringify(position)}`);
    await this.sleep(2000);
  }

  async giveItems(playerName, itemType, quantity) {
    const amount = quantity || this.inventory.get(itemType) || 0;
    
    if (amount === 0) {
      this.sendMessage(`Desculpe, não tenho ${itemType} no meu inventário.`);
      return;
    }

    console.log(`Entregando ${amount}x ${itemType} para ${playerName}`);
    
    this.removeFromInventory(itemType, amount);
    this.sendMessage(`Entregando ${amount}x ${itemType}!`);
  }

  addToInventory(item, quantity) {
    const current = this.inventory.get(item) || 0;
    this.inventory.set(item, current + quantity);
  }

  removeFromInventory(item, quantity) {
    const current = this.inventory.get(item) || 0;
    this.inventory.set(item, Math.max(0, current - quantity));
  }

  sendMessage(message) {
    try {
      if (this.bot.queue && this.bot.queue.length < 10) {
        this.bot.queue('text', {
          type: 'chat',
          needs_translation: false,
          source_name: this.bot.username,
          xuid: '',
          platform_chat_id: '',
          message: message
        });
      }
      console.log(`[BOT] ${message}`);
    } catch (error) {
      console.error('Erro ao enviar mensagem:', error);
    }
  }

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
